src/flash.rs

```diff
@@ -0,0 +1,254 @@
+    fn is_unlocked(&self) -> bool {
+        self.cr.read().lock().bit_is_clear()
+    }
+
+    /// Erases a page of Flash memory
+    ///
+    /// Attention:
+    /// Flash must be unlocked with `unlock` before `erase_page` can be used.
+    /// Make sure that your program is not executed from the
+    /// same Flash bank that you are going to erase.
+    fn erase_page(&mut self, page: u32) -> Result<()> {
+        verify_unlocked()?;
+        verify_page(page)?;
+
+        // Wait, while the memory interface is busy.
+        while self.sr.read().bsy().bit_is_set() {}
+
+        reset_errors();
+        self.cr.modify(|_, w| unsafe {
+            w
+            .per().set_bit()
+            .pnb().bits(page as u8)
+            .strt().set_bit()
+        });
+
+        // Wait for operation to complete
+        while self.sr.read().bsy().bit_is_set() {}
+        self.cr.modify(|_, w| w.per().clear_bit());
+
+        check_errors()
+    }
+
+    /// Writes a double word to Flash memory
+    ///
+    /// Attention:
+    /// Flash must be unlocked with `unlock` before `write_double_word` can be used.
+    /// Page where address points to must be erased using `erase_page` before writing.
+    /// If you use this method to write to Flash memory, the address must have
+    /// been erased before, otherwise this method will return an error.
+    fn write_double_word(&mut self, address: u32, word: u64) -> Result<()> {
+        verify_unlocked()?;
+        verify_address(address)?;
+
+        // Wait, while the memory interface is busy.
+        while self.sr.read().bsy().bit_is_set() {}
+
+        // Enable Flash programming
+        reset_errors();
+        self.cr.modify(|_,w| w.pg().set_bit());
+
+        // It is only possible to program a double word (2 x 32-bit data).
+        let address = address as *mut u32;
+
+        // We absoluty can't have any access to Flash while preparing the
+        // write, or the process will be interrupted. This includes any
+        // access to the vector table or interrupt handlers that might be
+        // caused by an interrupt.
+        interrupt::free(|_| {
+            // Safe, because we've verified the valididty of `address`.
+            unsafe {
+                address.write_volatile(word as u32);
+                address.offset(1).write_volatile((word >> 32) as u32);
+            }
+        });
+
+        // Wait for operation to complete
+        while self.sr.read().bsy().bit_is_set() {}
+        self.cr.modify(|_, w| w.pg().clear_bit());
+        check_errors()
+    }
+
+    /// Read from flash.
+    /// Returns a &[u8] if the address and length are valid.
+    /// Length must be a multiple of 4.
+    fn read(&mut self, address: u32, length: usize) -> Result<&'static [u8]> {
+        verify_address(address)?;
+        verify_length(length)?;
+        let address = address as *const _;
+        unsafe {
+            Ok(core::slice::from_raw_parts::<'static, u8>(address, length))
+        }
+    }
```

Comment:
Is it actually sound?
```rust
let slice = f.read(addr, 8);
f.write(addr, data);
```
After that slice points to new data. Is it ok, or should we come up with somekind of reader writer locking mechanism.

Reply:
It should be handled here for F1 flash with read/write grants: https://github.com/stm32-rs/stm32f1xx-hal/pull/141/files